Form, views, tests, some fixes to other places

adp-atea opened the pull request.

adp-atea committed:
Form, views, tests, some fixes to other places

sirex reviewed (changes requested):
https://github.com/atviriduomenys/katalogas/pull/162#pullrequestreview-1123703725

adp-atea committed:
Merge branch 'devel' into 154-šaltinio-forma

adp-atea committed:
Merge branch 'devel' into 154-šaltinio-forma

adp-atea committed:
marged from devel to fix migration issue

sirex reviewed (changes requested):
I did several changes, to fix the distribution form, I think this way we can move faster. But my changes are not complete, so I would like to ask you to finish with refactoring of some places, adding schema and data migrations.

Where is a list of things that needs to be done:

- [x] `DistributionFormat.url_format` was renamed to `format`. Previously it was used only for URL type distributions, but now it is needed for both URL and FILE distributions. Please finish renaming this field. And also add data migration, by selecting a `Format` instance for `type=FILE`. To do that, you need to detect correct format from `mime_type` and `filename` extension. You can find and example, how I was doing it [here](https://nbviewer.org/gist/sirex/90b291c559bafed8295f61606ea4fa6a#Brandos-lygis).
- [x] `DistributionFormat.url` was renamed to `download_url`, to match DCAT naming and differentiate from `access_url`, which was introduced quite recently. Please finish the renaming.
- [x] `DistributionFormat.filename` was renamed to `file`, since this is no longer just a file name, but a `File` object.
- [x] `DistributionFormat` `.region` and `.municipality` types changed from `CharField` to `ForeignKey`, please finish refactoring other places where this was used and also add a schema and data migration. For data migration, you need to replace `region` and `municipality` values from titles to primary keys.
- [x] I think `DistributionFormat` fields `type` and `mime_type` are no longer needed, but do not delete them, just move down the field list and add a comment `Deprecated` above those fields.
- [x] Added [crispy_bulma](https://github.com/ckrybus/crispy-bulma) package in order to have a proper form styling, previous implementation was a mess. I did refactored `DatasetResourceForm` to accommodate this change, but could you please do the same thing to all other forms, where `FormHelper` is used?

I know, that is a lot of additional changes, but unfortunately, `DistributionFormat` was a painful places, and it needs to be fixed.

Since these are additional changes, I'm raising existing estimate from 12 to 24 hours.

sirex committed:
Fix forms styling and do some model fields refactoring

adp-atea committed:
Updated dependencies, added area in DatasetDistribution model where deprecated fields are placed

adp-atea committed:
Merge branch 'devel' into 154-šaltinio-forma

# Conflicts:
#	poetry.lock
#	pyproject.toml
#	vitrina/helpers.py

adp-atea committed:
Implemented crispy bulma to all forms using FormHelper

adp-atea committed:
Merge branch 'devel' into 154-šaltinio-forma

adp-atea committed:
DB migrations and data migrations for most new fields

adp-atea committed:
Migration

adp-atea committed:
Merge branch '154-šaltinio-forma' of https://github.com/atviriduomenys/katalogas into 154-šaltinio-forma

adp-atea committed:
Merge branch 'devel' into 154-šaltinio-forma

adp-atea committed:
Split migration into many, for debugging purposes

adp-atea committed:
Removed region and municipality, replaced with geo_location field, edited field and data migration accordingly

adp-atea committed:
Finished renaming all fields that were changed, edited form to resemble original one

adp-atea committed:
Final fixes for tests, factories, views

adp-atea committed:
Merge branch 'devel' into 154-šaltinio-forma

# Conflicts:
#	poetry.lock
#	vitrina/datasets/models.py
#	vitrina/datasets/views.py
#	vitrina/helpers.py
#	vitrina/resources/models.py
#	vitrina/settings.py

adp-atea committed:
Fixes after merge

adp-atea committed:
poetry update

adp-atea committed:
Crispy forms downgrade to make bulma work

adp-atea committed:
Fixed crispy bulma

adp-atea committed:
Poetry update

adp-atea committed:
Specified elasticcsearch version, removed slug from dataset factory

sirex committed:
Cosmetic form readjustments

sirex commented on vitrina/resources/models.py:
```diff
@@ -40,70 +43,125 @@ class DatasetDistribution(models.Model):
+    geo_location = models.CharField(
+        max_length=255,
+        blank=True,
+        null=True,
```
```suggestion
```

Char fields should never be null, unless there is a good reason for that. Please change this also for `period_start` and `period_end`.

Also, when adding a form with empty `geo_location`, `period_start` or `period_end` I get this error:

```python
Traceback (most recent call last):
  File "django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.NotNullViolation: null value in column "geo_location" of relation "dataset_distribution" violates not-null constraint
DETAIL:  Failing row contains (10938, 2022-10-10 18:15:26.796594+00, 2022-10-10 18:15:26.796629+00, 1, , null, , 
null, null, null, null, null, null, , URL, http://example.com/data.csv, 1254, null, null, null, 1, , null).
```

It looks, that model says null=True, but migrations things it's null=False. Anyway, CharField, should almost never be null=True. Empty string fully satisfies emptiness, there is no need for special kind of emptiness for strings.

sirex commented on vitrina/resources/migrations/0007_files_fix.py:
```diff
@@ -0,0 +1,56 @@
+            detected_format = Format.objects.filter(mimetype=mime_type).first()
+            resource.format = detected_format
+        if resource.type == 'URL':
+            url_format = Format.objects.filter(extension='URL').first()
```
For URLs you actually need to do a real request and get format from response content type, like in example here:

```python
    status_code = 0
    content_type = ''
    try:
        resp = requests.get(url, stream=True, timeout=1)
    except requests.RequestException:
        pass
    else:
        if resp.status_code == 200:
            content_type = resp.headers.get('content-type')
            if ';' in content_type:
                content_type = content_type.split(';', 1)[0]
```

The you need to map received content type with a format:

```python
{
    'application/rdf+xml': 'RDF',
    'text/csv': 'CSV',
    'application/json': 'JSON',
    'application/geo+json': 'JSON',
    'application/xml': 'XML',
    'text/xml': 'XML',
    'application/vnd.ms-excel': 'XLSX',
    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet': 'XLSX',
    'text/html': 'HTML',
    'text/plain': 'URL',
    'application/zip': 'URL',
    '': 'URL',
}
```

And only fallback to URL if there is no way, to detect which format it actually is.

sirex replied:
By the way, I tink it is better to move this migration into a script, because this migration is most likely will take a long time to complete.

adp-atea committed:
Merge branch 'devel' into 154-šaltinio-forma

# Conflicts:
#	poetry.lock
#	pyproject.toml

adp-atea committed:
Modified migrations for formats to specify exact mimetypes. Made other fixes

adp-atea committed:
Fixed factory and tests to include now required date fields

adp-atea committed:
Merge branch 'devel' into 154-šaltinio-forma

# Conflicts:
#	vitrina/datasets/views.py
#	vitrina/orgs/forms.py
#	vitrina/resources/models.py

adp-atea committed:
Merged, fixed tests and views

adp-atea committed:
poetry fix

adp-atea committed:
Merge branch 'devel' into 154-šaltinio-forma

sirex committed:
Refactor format migration and other fixes

- Format migration was not correct, most distributions was of URL format
  and very few datasets had format set.

- Changed back  dist period to nullable field and change type from char
  to date as it should be.

- Fixed format line in dataset list.

- Added urlizetrunc filter to dataset description i details page.

- Removed links to Facebook, no one uses it anyway ant link occupies
  very important place.

adp-atea committed:
Merge branch 'devel' into 154-šaltinio-forma

# Conflicts:
#	poetry.lock
#	pyproject.toml
#	vitrina/datasets/factories.py
#	vitrina/datasets/views.py
#	vitrina/settings.py

adp-atea committed:
Minor fixes after merge

adp-atea committed:
Poetry fix
